app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+
+
+        var doesFileParentMatch = false
+        val fileParent = file.parentFile
+        if (fileParent == null || fileParent.equals(parent)) {
+            doesFileParentMatch = true
+        }
+
+        if (parent.equals(file) || doesFileParentMatch) {
+            if (file.delete()) {
+
+            } else {
+                throw FileNotFoundException("Failed to delete document with id $documentId")
+            }
+        } else {
+            throw FileNotFoundException("Failed to delete document with id $documentId")
+        }
+    }
+
+    override fun renameDocument(documentId : String?, displayName : String?) : String? {
+
+        if (displayName == null) {
+            throw FileNotFoundException("Couldn't rename document, new name is null")
+        }
+
+        val sourceFile = getFileForDocId(documentId!!)
+        val sourceParentFile = sourceFile.parentFile ?: throw FileNotFoundException("Failed to rename document. File has no parent.")
+        val destFile = File(sourceParentFile.path, displayName)
+
+        try {
+            val renameSucceeded = sourceFile.renameTo(destFile)
+            if (!renameSucceeded) {
+                throw FileNotFoundException("Failed to rename document. Renamed failed.")
+            }
+        } catch (e : Exception) {
+            throw FileNotFoundException("Couldn't rename document. Error: " + e.message)
+        }
+        return getDocIdForFile(destFile).toString()
+    }
+
+    private fun copyDocument(
+        sourceDocumentId : String, sourceParentDocumentId : String,
+        targetParentDocumentId : String?
+    ) : String? {
+        if (!isChildDocument(sourceParentDocumentId, sourceDocumentId)) {
+            throw FileNotFoundException(
+                "Failed to copy document with id " +
+                        sourceDocumentId + ". Parent is not: " + sourceParentDocumentId
+            )
+        }
+        return copyDocument(sourceDocumentId, targetParentDocumentId)
+    }
+
+    override fun copyDocument(sourceDocumentId : String, targetParentDocumentId : String?) : String? {
+
+        val parent = getFileForDocId(targetParentDocumentId!!)
+        val oldFile = getFileForDocId(sourceDocumentId)
+        val newFile = File(parent.path, oldFile.name)
+        try {
+
+            var wasNewFileCreated = false
+            if (newFile.createNewFile()) {
+                if (newFile.setWritable(true) && newFile.setReadable(true)) {
+                    wasNewFileCreated = true
+                }
+            }
+            if (!wasNewFileCreated) {
+                throw FileNotFoundException(
+                    "Couldn't copy document " + sourceDocumentId +
+                            ". Could not create new file."
+                )
+            }
+            FileInputStream(oldFile).use { inStream ->
+                FileOutputStream(newFile).use { outStream ->
+                    val buf : ByteArray = ByteArray(4096)
```

Comment:
You don't need `: ByteArray` if you're assigning a `ByteArray` to it, additionally there should be a comment here denoting that you're copying 4 KiB of data at a time. Albeit, I do find it concerning that you need to do a copy in this manner, does `File` not have a method like `File#copy`?